design/Events_design.md

```diff
@@ -0,0 +1,113 @@
+- stream all "transfer" events where the 'to' field is my address
+- stream all "transfer" events from the USDC contract
+
+
+### How events work 
+
+The query made on the UI side is transformed in a server-side query on the node, and there is some logic after a tx is executed and events are emmitted, to match them against the filters requested by users and distribute them to the requester.(? - need to confirm this is 100% how it works) .
+
+*Note that there is no constraint on data access, since all data is public.*
+
+
+## Requirements and Problems in Obscuro
+
+- The high level requirement is to determine which account owner can view which events, and make sure, through encryption that nobody else can see those events.
+- Ideally we want to achieve this without any change to the API.
+
+
+#### Problem1: How do we authenticate event subscriptions? 
+We have a wallet extension with viewing key support, which we have to use ideally, to avoid creating another mechanism.
+
+#### Problem2: Assuming the enclave knows who is subscribed, how will the filtering work?
+
+E.g: Alice is requesting to listen to all transfer events. (this is a perfectly reasonable request on a public chain)
+
+There are 2 cases:
+1. The event was emitted by a smart contract as a result of executing a tx sent by Alice.
+2. The event was emitted as result of a Tx sent by Bob but it is somehow relevant to Alice.
+
+
+#### Problem 3: How to encrypt events before they leave the enclave
+
+#### Problem 4: The mechanism should not leak general information on what activity is happening.
+
+
+## Proposed solution
+
+### Prerequisites
+An event `E1` is emitted after executing transaction `TX1` with a from address `A1`.
+`E1` is formed of multiple topics, 2 of which are addresses (`A2` and `A3`), and there are some other random values.
+
+An event `E2` is emitted by a transaction `TX2` from an address `A4`.
+`E2` is formed of multiple topics, none of which are addresses.
+
+
+### Solution
+
+1. We modify the subscription call to accept a list of owning accounts. 
+Each account must be signed with the latest viewing key, to prevent someone from asking random events, just to leak info. 
+The call will fail if there are no viewing keys for all those accounts.
+This is possible because the subscription call is implemented on Obscuro, and made by the wallet_extension, so it doesn't have to be compatible with Ethereum.
+For our RPC client it would an authenticated subscription.
+
+2. We need to pass into the enclave the "Subscription" objects, and create an rpc streaming for each. 
+
+3. The enclave runs all the usual filtering logic on this event, and determines if there is any subscription that requested it.
+
+4. Then, there is an extra step (inside the enclave as well) to determine whether that party is allowed to view that event. 
+
+This can be done by running an intersection between:
+  - all accounts found in the event topics plus the `from` of the transaction: `E1->(A1,A2,A3)` 
```

Comment:
Will it always be clear which topics are addresses vs not addresses? Is it possible some topics will be hashes (this seemed common based on the link you provided, Tudor) that will look address-like?